File: framework/test2/floating/matrix/matrix.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.leetcode.nest import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass
class Solution:
    def matrixMultiply(self, a: List[List[float]], b: List[List[float]]) -> List[List[float]]:
        rows = len(a)
        cols = len(b[0])
        res = [[0.0] * cols for i in range(rows)]
        for i in range(rows):
            for j in range(cols):
                c = 0.0
                for k in range(len(b)):
                    c += a[i][k] * b[k][j]
                res[i][j] = c
        return res

if __name__ == '__main__':
    init_logging()
    from soda.unittest import Validators
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().matrixMultiply)
    # OR use struct tester
    # work = TestWork.forStruct(CLASS)

    # step [3]: setup other options
    # work.validator = (e,r) => bool
    # work.compareSerial = True
    work.run()

File: framework/test2/floating/matrix/matrix.rb
# step [1]: implement solution function
# @param {Float[][]} a
# @param {Float[][]} b
# @return {Float[][]}
def matrix_multiply(a, b)
  rows, cols = a.size, b[0].size
  res = Array.new(rows){Array.new(cols, 0.0)}
  (0...rows).each { |i|
    (0...cols).each { |j|
      c = 0.0
      (0...b.size).each { |k|
        c += a[i][k] * b[k][j]
      }
      res[i][j] = c
    }
  }
  res
end

if __FILE__ == $0
  require 'soda/unittest/work'
  ns = Soda::Unittest
  # step [2]: setup function/return/arguments
  work = ns::TestWork.create(method(:matrix_multiply))
  # work = ns::TestWork.forStruct(CLASS)
  # work.validator = -> (e, r) { e is equal to r }
  # work.compare_serial = true
  puts work.run(ARGF.read)
end

